### PR TITLE
Fluent UI v2 CSS provider

### LINKED ISSUE #4135: Implement Microsoft Fluent UI
Blazorise supports all the major themes, which is very nice!

But since this is a blazor library - a Microsoft product - it would be nice to also have a Microsoft theme as well. Then we can make apps that look like azure / office365 / windows / etc.

That design system is called "FAST", and the theme is called "FluentUI". There is already a [blazor library](https://github.com/microsoft/fast-blazor) for that, but it is small and lacks all the features in blazorise - importantly, it does not have layouts, grids, validations, breakpoints, css utilities, etc., etc., etc. It only has components.

### LINKED ISSUE #5287: Update Font Awesome version to 6.x
update font awesome version (currently 5.15.4) to 6.5.1.

would increase the number of free icons - from 1600 to 2000.

### LINKED ISSUE #5292: Rating: add tooltip for the items
This will make it easy for the user to indicate the right score

**Example**

image

### PR DESCRIPTION
**Pull request by stsrki**
Closes #4135
Closes #5287
Closes #5292

### CODE REVIEW

**Comment by stsrki** [comment]
@David-Moreira please review once you get the chance.

**Comment by David-Moreira** [comment]
> @David-Moreira please review once you get the chance.

image

LGTM...?? hahaha

**Comment by stsrki** [comment]
> > @David-Moreira please review once you get the chance.
> 
> image
> 
> LGTM...?? hahaha

Pfff, that's nothing 🤣

**Comment by David-Moreira** [comment]
**UI Review based on our Demo:**

**Disclaimer :** I have never ever used Fluent UI neither am well acquainted with it, so I can't comment whether something looks like it should in Fluent UI. I will comment things that I think might be wrong or done better.

1. Header should stay on top, the body also seems to be cut off, it's behind the header
image

2. Theme settings seem to not have any effect?
image

3. The mouse hover here, does not display a clickable indicator, is this expected?
image

4. Is it expected that the alerts are hidden instead of removed from the html? For example, bs seems to regain back the space once the alerts are closed.
image
image

5. Do the menu items not have a selected / active look?
image

6. Does fluent UI use explicit text for previous/next? Does this needs to be localized?
image

7. The Modal Animation for Fade In does not seem to obbey animation duration

8. The pagination has no hover look or anything, seems very static, is this expected? Also it should be visible which pages are selected/active.

9. The Danger Page Progress does not seem to be different.

10. Table Fixed Columns seems to be broken
image

11. Should Addon be this big?
image

12. Table Resize feature seems to be broken, both Header And Columns

13. The Faded tooltip looks like the others?

14. The naming should actually be Start/End instead of Left/Right?
image

15. Should the 100 represent the currently selected value, or the maximum value?
image

16. No radio buttons
image

17. No color at all
image

18. Is this how addons look? I can't really tell the Start apart from the text. Also not enough spacing? Is this how fluent UI states it should be like?
image

19. I guess a Dropdown might not be natively supported as addon, but maybe it should not have the left padding?
image

20. Addon button right padding?
image

21. Some addons don't look very good
image

22. Why does the DataGrid icon look like a music note?
image

23. Pagination in the DataGrid : 
- I need to know which page is selected
- Previous/Next First/Last buttons should appear as disabled accordingly
- The dropdown page selection should only appear in certain resolutions?
image

24. Datagrid Column filters don't look too good
image

25. DataGrid Fixed Header too big?
image

26. Pagination look. Some icons don't seem correct?
image

27. DataGrid Resize mode, also broken (table related) http://localhost:5138/tests/datagrid/resize

28. Fixed Columns create a white background on top of selected look
image

29. DataGrid Multiple : Clicking the checkbox does not check it
image

30. Datagrid Popup Edit with horizontal bar (overflow)
image

31: DataGrid Header looks too big (It's from resize feature being activated apparently)
image

32. Autocomplete selection does not look very good
image

33. I believe the dropdown list has max height or something. There should be a scroll
image

34. The top snackbars appear behind the header
http://localhost:5138/tests/snackbar

35. Are all the borders correct?
image

**Comment by stsrki** [comment]
Work so far

1. fixes
2. I have planned for this. We will skip it for noe.
3. fixed
4. fixed
5. fixed
6. working as expected, and we also have localization
7. 
8. fixed
9. fixed
10. David
11. demo ui adjusted
12. David
13. fixed
14. fixed
15. yes, the value shows min and max acording to fluent design
16. 
17. 
18. buttons inside of addons asjusted to not have colors, acording to fluent design
19. save as 18. Padding must be there acording to fluent design
20. same as 18. and 19.
21. as I don't have a reference for diferent addon sizes I had to improvise. Until we know more this is how it works.
22. fixed. introduces new Table icon
23. fixed
24. 
25. fixed by adding missing CSS
26. these icons are the best I could find in the Fluent UI icons package. If you want, you can look at https://react.fluentui.dev/?path=/docs/concepts-developer-icons-icons-catalog--page for alternative
27. fixed with 25.
28. 
29. 
30. 
31. fixed by adding missing CSS
32. David
33. fixed
34. 
35.

File: Source/Blazorise.FluentUI2/Styles/components/_card.scss [hunk 1]
```diff
@@ -5,8 +5,6 @@
     position: relative;
     display: flex;
     flex-direction: column;
-    overflow-x: hidden;
-    overflow-y: hidden;
```
Comment (David-Moreira) [inline Source/Blazorise.FluentUI2/Styles/components/_card.scss]: @stsrki Please evaluate this change. 
The problem is if the overflow is not permitted, inner content will not be available to overflow on top of the card.

Comment (David-Moreira) [reply Source/Blazorise.FluentUI2/Styles/components/_card.scss]: (Autocomple dropdown fix)

Comment (stsrki) [reply Source/Blazorise.FluentUI2/Styles/components/_card.scss]: What is the problem here? Can you clarify?

Comment (David-Moreira) [reply Source/Blazorise.FluentUI2/Styles/components/_card.scss]: Did I not just say? 
image

Comment (stsrki) [reply Source/Blazorise.FluentUI2/Styles/components/_card.scss]: Yes, but I tried it on my side, and it works. That's why I'm asking for more information.

Comment (David-Moreira) [reply Source/Blazorise.FluentUI2/Styles/components/_card.scss]: It works due to this change

Comment (stsrki) [reply Source/Blazorise.FluentUI2/Styles/components/_card.scss]: It seems it is not working after all.

image

I will look into it.

Comment (David-Moreira) [reply Source/Blazorise.FluentUI2/Styles/components/_card.scss]: Ok, so I fixed Autocomplete, and I broke that one then?
Please test Autocomplete with whatever changes you make.

Comment (stsrki) [reply Source/Blazorise.FluentUI2/Styles/components/_card.scss]: It should be fixed now. Autocomplete also works.

**Comment by stsrki** [comment]
On Autocomplete, we still show checked items with an icon.

image

Now that we have real checkboxes on DropdownItem it would look much better with real checkboxes.

image

image

---

@David-Moreira Would you like to work on it? In this PR or should we open new ticket once we merge this Fluent UI?

**Comment by stsrki** [comment]
Leaving what was left to do

**UI Review based on our Demo:**

10. Table Fixed Columns seems to be broken
image

24. Datagrid Column filters don't look too good
image

29. DataGrid Multiple : Clicking the checkbox does not check it
image

**Comment by David-Moreira** [comment]
**12.** Fixed 

Can you do **11**? You probably better suited since it's styling. And you're already used to the fluent styling?

File: Source/Extensions/Blazorise.DataGrid/_DataGridPagination.razor [hunk 2]
```diff
@@ -175,7 +175,7 @@
                     var pageNumberString = i.ToString();
                     var pageActive = pageNumber == PaginationContext.CurrentPage;
 
-                    <PaginationItem Display="Blazorise.Display.None.OnMobile.InlineBlock.OnTablet" Disabled="@(pageNumber == PaginationContext.CurrentPage)" Active="@pageActive">
+                    <PaginationItem Display="Blazorise.Display.None.OnMobile.InlineFlex.Row.OnTablet" Disabled="@(pageNumber == PaginationContext.CurrentPage)" Active="@pageActive">
```
Comment (David-Moreira) [inline Source/Extensions/Blazorise.DataGrid/_DataGridPagination.razor:178]: This is a wide change, is this ok for all providers?

Comment (stsrki) [reply Source/Extensions/Blazorise.DataGrid/_DataGridPagination.razor:178]: Yes, I tried in all providers. Works without visible issues.

**Comment by stsrki** [comment]
> **12.** Fixed
> 
> Can you do **11**? You probably better suited since it's styling. And you're already used to the fluent styling?

Much better, I think

image

**Comment by David-Moreira** [comment]
> > **12.** Fixed
> > Can you do **11**? You probably better suited since it's styling. And you're already used to the fluent styling?
> 
> Much better, I think
> 
> image

Yep, you the man!!

**Comment by stsrki** [comment]
I'm eager to merge. Wanna QA fast and not tell me a thing? 🤣🤣

**Comment by David-Moreira** [comment]
1. I still see no visible changes when trying to change the theme?
image

2. Is this the expected Modal look? Doesn't look like the header & footer are noticeably separate from the body.

image

3. The pagination has no hover look, usually clickable / interactible items have some visual hint to them. Does fluent ui not have that? Regular buttons seem to have it.

4. I don't see anything in the 2 pixels?

image

All in all, everything acts and looks better to me.

File: Demos/Blazorise.Demo.FluentUI2/Blazorise.Demo.FluentUI2.csproj [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  </ItemGroup>
+
+  <ItemGroup>
+    <Content Update="_Imports.razor">
+      <ExcludeFromSingleFile>true</ExcludeFromSingleFile>
+      <Generator>MSBuild:RazorGenerateComponentDeclarationDesignTime</Generator>
+    </Content>
+  </ItemGroup>
+
+</Project>
```
Comment (David-Moreira) [inline Demos/Blazorise.Demo.FluentUI2/Blazorise.Demo.FluentUI2.csproj:22]: Should this be here?

File: Demos/Blazorise.Demo.FluentUI2/dotnetwatchrun.bat [hunk 4]
```diff
@@ -0,0 +1,6 @@
+: This is here to enable console colors when using watch
+: Usefull for debugging when developping
+set DOTNET_WATCH_SUPPRESS_LAUNCH_BROWSER=1
+title Blazorise.Demo.Bulma
```
Comment (David-Moreira) [inline Demos/Blazorise.Demo.FluentUI2/dotnetwatchrun.bat:4]: wrong title

File: Source/Blazorise.FluentUI2/Components/AlertMessage.razor [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+            else if ( ParentAlert?.Color == Color.Warning )
+            {
+                <svg fill="currentColor" aria-hidden="true" width="1em" height="1em" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg"><path d="M8.68 2.79a1.5 1.5 0 0 1 2.64 0l6.5 12A1.5 1.5 0 0 1 16.5 17h-13a1.5 1.5 0 0 1-1.32-2.21l6.5-12ZM10.5 7.5a.5.5 0 0 0-1 0v4a.5.5 0 0 0 1 0v-4Zm.25 6.25a.75.75 0 1 0-1.5 0 .75.75 0 0 0 1.5 0Z" fill="currentColor"></path></svg>
+            }
+            else 
+            {
+                <svg fill="currentColor" aria-hidden="true" width="1em" height="1em" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg"><path d="M18 10a8 8 0 1 0-16 0 8 8 0 0 0 16 0ZM9.5 8.91a.5.5 0 0 1 1 0V13.6a.5.5 0 0 1-1 0V8.9Zm-.25-2.16a.75.75 0 1 1 1.5 0 .75.75 0 0 1-1.5 0Z" fill="currentColor"></path></svg>
+            }
+        </div>
+    } *@
```
Comment (David-Moreira) [inline Source/Blazorise.FluentUI2/Components/AlertMessage.razor:23]: Is this all commented?

**Review by David-Moreira — APPROVED** [review]
LGTM, most of the files I glanced over. Too many files...

**Comment by stsrki** [comment]
1. as I have mentioned before. For now I'm not gonna do the theme generator. As an alternative, users could use Fluent design tokens, which I will document before release.
2. Yes, this is how it looks natively. In fact it works even worse. In our version we at least have some spacings.
3. Fixed it
4. This works, but since Secondary color has a low contrast compared to background it is not so much visible.

**Comment by stsrki** [comment]
Regarding the theme generator. Basically, this is my idea:

- Goto https://react.fluentui.dev/?path=/docs/theme-theme-designer--page
- Configure theme
- Export
- Copy tokens
- Apply tokens in Blazorise somehow
- Generate CSS variables from JSON tokens

image

---

That is how it is done manually.

We can expand from there and automate it a little. Generate a C# model from the JSON object with all of the token fields. Then, in Blazorise, the theme generator applies colors and other variables to regenerate new JSON object, and in turn, it generates CSS new variables.

**Comment by David-Moreira** [comment]
I see ok than. I guess we can improve that later.

**Pull request merged**